File: Components/Document Server/VB/DocumentServerWebDemo/App_Code/SpreadsheetAPI/XLExportExampleCodeEvaluator.vb
Imports Microsoft.VisualBasic
Imports DevExpress.Export.Xl
Imports System
Imports System.CodeDom.Compiler
Imports System.Reflection

Namespace DevExpress.DXperience.Demos
	#Region "CodeEvaluationResult"
	Public Enum CodeEvaluationResult
		Success
		CompileErrors
		ExecuteErrors
	End Enum
	#End Region
	#Region "XLExportExampleCodeEvaluatorBase"
	Public MustInherit Class XLExportExampleCodeEvaluatorBase
		Protected MustOverride ReadOnly Property CodeStart() As String
		Protected MustOverride ReadOnly Property CodeEnd() As String
		Protected MustOverride Function GetCodeDomProvider() As CodeDomProvider
		Protected Function GetModuleAssembly() As String
			Return AssemblyInfo.SRAssemblySpreadsheetCore
		End Function
		Protected Function GetExampleClassName() As String
			Return "XLExportExampleCode.ExampleItem"
		End Function

		Public Function ExecuteCodeAndGenerateDocument(ByVal exampleCode As String, ByVal parameters() As Object) As CodeEvaluationResult
			Dim theCode As String = String.Concat(CodeStart, exampleCode, CodeEnd)
			Dim linesOfCode() As String = { theCode }
			Return CompileAndRun(linesOfCode, parameters)
		End Function

		Protected Friend Function CompileAndRun(ByVal linesOfCode() As String, ByVal parameters() As Object) As CodeEvaluationResult
			Dim CompilerParams As New CompilerParameters()

			CompilerParams.GenerateInMemory = True
			CompilerParams.TreatWarningsAsErrors = False
			CompilerParams.GenerateExecutable = False

			Dim referencesSystem() As String = { "System.dll", "System.Windows.Forms.dll", "System.Data.dll", "System.Xml.dll", "System.Drawing.dll" }

			Dim referencesDX() As String = { AssemblyInfo.SRAssemblyData, GetModuleAssembly(), AssemblyInfo.SRAssemblyOfficeCore, AssemblyInfo.SRAssemblyPrintingCore, AssemblyInfo.SRAssemblyPrinting, AssemblyInfo.SRAssemblyDocs, AssemblyInfo.SRAssemblyUtils }
			Dim references(referencesSystem.Length + referencesDX.Length - 1) As String

			For referenceIndex As Integer = 0 To referencesSystem.Length - 1
				references(referenceIndex) = referencesSystem(referenceIndex)
			Next referenceIndex

			Dim i As Integer = 0
			Dim initial As Integer = referencesSystem.Length
			Do While i < referencesDX.Length
				Dim [assembly] As System.Reflection.Assembly = System.Reflection.Assembly.Load(referencesDX(i) + AssemblyInfo.FullAssemblyVersionExtension)
				If [assembly] IsNot Nothing Then
					references(i + initial) = [assembly].Location
				End If
				i += 1
			Loop
			CompilerParams.ReferencedAssemblies.AddRange(references)


			Dim provider As CodeDomProvider = GetCodeDomProvider()
			Dim compile As CompilerResults = provider.CompileAssemblyFromSource(CompilerParams, linesOfCode)

			If compile.Errors.HasErrors Then
				Return CodeEvaluationResult.CompileErrors
			End If

			Dim [module] As System.Reflection.Module = Nothing
			Try
				[module] = compile.CompiledAssembly.GetModules()(0)
			Catch
			End Try

			Dim moduleType As Type = Nothing

			If [module] Is Nothing Then
				Return CodeEvaluationResult.CompileErrors
			End If

			moduleType = [module].GetType(GetExampleClassName())

			Dim methInfo As MethodInfo = Nothing
			If moduleType Is Nothing Then
				Return CodeEvaluationResult.CompileErrors
			End If

			methInfo = moduleType.GetMethod("Process")

			If methInfo Is Nothing Then
				Return CodeEvaluationResult.CompileErrors
			End If
			Try
				methInfo.Invoke(Nothing, parameters)
			Catch ex As Exception
				ShowErrorMessage(ex.InnerException.Message)
				Return CodeEvaluationResult.ExecuteErrors
			End Try
			Return CodeEvaluationResult.Success
		End Function
		Protected Overridable Sub ShowErrorMessage(ByVal message As String)
		End Sub
	End Class
	#End Region
End Namespace

File: Components/Document Server/CS/DocsMainDemo/Modules/Api/XLExportExampleCodeEvaluator.cs
using DevExpress.Export.Xl;
using System;
using System.CodeDom.Compiler;
using System.Reflection;

namespace DevExpress.DXperience.Demos {
    #region CodeEvaluationResult
    public enum CodeEvaluationResult {
        Success,
        CompileErrors,
        ExecuteErrors
    }
    #endregion
    #region XLExportExampleCodeEvaluatorBase
    public abstract class XLExportExampleCodeEvaluatorBase {
        protected abstract string CodeStart { get; }
        protected abstract string CodeEnd { get; }
        protected abstract CodeDomProvider GetCodeDomProvider();
        protected string GetModuleAssembly() {
            return AssemblyInfo.SRAssemblySpreadsheetCore;
        }
        protected string GetExampleClassName() {
            return "XLExportExampleCode.ExampleItem";
        }

        public CodeEvaluationResult ExecuteCodeAndGenerateDocument(string exampleCode, object[] parameters) {
            string theCode = String.Concat(CodeStart, exampleCode, CodeEnd);
            string[] linesOfCode = new string[] { theCode };
            return CompileAndRun(linesOfCode, parameters);
        }

        protected internal CodeEvaluationResult CompileAndRun(string[] linesOfCode, object[] parameters) {
            CompilerParameters CompilerParams = new CompilerParameters();

            CompilerParams.GenerateInMemory = true;
            CompilerParams.TreatWarningsAsErrors = false;
            CompilerParams.GenerateExecutable = false;

            string[] referencesSystem = new string[] { "System.dll",
                                                      "System.Windows.Forms.dll",
                                                      "System.Data.dll",
                                                      "System.Xml.dll",
                                                      "System.Drawing.dll" };

            string[] referencesDX = new string[] {
                AssemblyInfo.SRAssemblyData,
                GetModuleAssembly(),
                AssemblyInfo.SRAssemblyOfficeCore,
                AssemblyInfo.SRAssemblyPrintingCore,
                AssemblyInfo.SRAssemblyPrinting,
                AssemblyInfo.SRAssemblyDocs,
                AssemblyInfo.SRAssemblyUtils
            };
            string[] references = new string[referencesSystem.Length + referencesDX.Length];

            for (int referenceIndex = 0; referenceIndex < referencesSystem.Length; referenceIndex++) {
                references[referenceIndex] = referencesSystem[referenceIndex];
            }

            for (int i = 0, initial = referencesSystem.Length; i < referencesDX.Length; i++) {
                Assembly assembly = Assembly.Load(referencesDX[i] + AssemblyInfo.FullAssemblyVersionExtension);
                if (assembly != null)
                    references[i + initial] = assembly.Location;
            }
            CompilerParams.ReferencedAssemblies.AddRange(references);


            CodeDomProvider provider = GetCodeDomProvider();
            CompilerResults compile = provider.CompileAssemblyFromSource(CompilerParams, linesOfCode);

            if (compile.Errors.HasErrors)
                return CodeEvaluationResult.CompileErrors;

            Module module = null;
            try {
                module = compile.CompiledAssembly.GetModules()[0];
            }
            catch { }

            Type moduleType = null;

            if (module == null)
                return CodeEvaluationResult.CompileErrors;

            moduleType = module.GetType(GetExampleClassName());

            MethodInfo methInfo = null;
            if (moduleType == null)
                return CodeEvaluationResult.CompileErrors;

            methInfo = moduleType.GetMethod("Process");

            if (methInfo == null)
                return CodeEvaluationResult.CompileErrors;
            try {
                methInfo.Invoke(null, parameters);
            }
            catch (Exception ex) {
                ShowErrorMessage(ex.InnerException.Message);
                return CodeEvaluationResult.ExecuteErrors;
            }
            return CodeEvaluationResult.Success;
        }
        protected virtual void ShowErrorMessage(string message) {
        }
    }
    #endregion
}
